rest_docs/rest-catalog-open-api-v0.1.yaml

```diff
@@ -0,0 +1,727 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+---
+openapi: 3.0.3
+info:
+  title: Apache Iceberg REST Catalog API
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: 1.0.0
+  description:
+    Defines the specification for the first version of the REST Catalog API. Implementations should support both Iceberg table specs v1 and v2, with priority given to v2.
+servers:
+  - url: http://127.0.0.1:1080
+    description: URL Used for Mock-Server Unit Tests
+# All routes are currently configured using an Authorization header.
+security:
+  - BearerAuth: []
+paths:
+  /v1/config:
+    get:
+      tags:
+        - Configuration API
+      summary: List all catalog configuration settings
+      operationId: getConfig
+      description:
+        All REST catalogs will be initialized by calling this route. This route
+        will return at least the minimum necessary metadata to initialize the
+        catalog. Optionally, it can also include server-side specific overrides.
+        For example, it might also include information used to initialize this catalog
+        such as the details of the Http connection pooling, etc. This route might
+        also advertise information about operations that are not implemented
+        so that the catalog can eagerly throw or go about another way of performing
+        the desired action.
+      responses:
+        default:
+          description: Server-Specific Configuration Overrides
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/IcebergConfiguration'
+        "400":
+          description: Unknown Error
+        "401":
+          description: Unauthorized
+
+  /v1/namespaces/{namespace}/tables/{table}:
```

Comment:
My understanding is that, at least per the spec for URIs, that individual components should be encoded after their separation character.

I wonder if any server frameworks would choke on a path parameter that essentially allows for [0, n] / (0, n] values separated by a slash? Having `tables` there makes it consistent with some of the other routes and provides a fully separator.

Namespace will already have to be encoded if it includes special characters, so I'm partial to the way I have it now, but if others feel strongly about your proposed idea, I'm certainly open to it.

Reply:
I think @rymurr is asking why we include the fixed parts of the path, `namespaces` and `tables`.

That's because we want to avoid conflicts with routes like `/v1/namespaces/{ns}/properties`. We want to be able to have that route for properties, but also support tables named properties at `/v1/namespaces/{ns}/table/properties`.

Another option is to use different prefixes, like `/properties/{ns}`. But then you end up with a lot of prefixes, and I think that those tend to lack the helpful context. What does `/properties/accounting/q3_2021` return? Information about a table or a namespace? It isn't obvious from the path, but a slightly more verbose `/namespaces/accounting/tables/q3_2021/properties` is clear. (Of course, you can add even more prefixes, like `/namespace_properties/{ns}` and `/table_properties/{ns}/{tab}`)

I like the approach of including the context and having the hierarchy there.